src/zx/FunctionalityConstruction.cpp

```diff
@@ -0,0 +1,344 @@
+                               Expression(PiRational(op->getParameter()[2])) + PiRational(1, 1));
+                    addXSpider(diag, target, qubit_vertices, Expression(PiRational(1, 2)));
+                    addZSpider(diag, target, qubit_vertices,
+                               Expression(PiRational(op->getParameter()[1])) + PiRational(3, 1));
+                    break;
+                }
+
+                case qc::OpType::SWAP: {
+                    const auto target2 = op->getTargets()[1];
+                    addSwap(diag, target, target2, qubit_vertices);
+                    break;
+                }
+                case qc::OpType::H: {
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    break;
+                }
+                case qc::OpType::Measure:
+                case qc::OpType::I: {
+                    break;
+                }
+                default: {
+                    throw ZXException("Unsupported Operation: " +
+                                      qc::toString(op->getType()));
+                }
+            }
+        } else if (op->getNcontrols() == 1 && op->getNtargets() == 1) {
+            const auto target = op->getTargets().front();
+            const auto ctrl   = (*op->getControls().begin()).qubit;
+            switch (op->getType()) { // TODO: any gate can be controlled
+                case qc::OpType::X: {
+                    // check if swap
+                    if (checkSwap(it, end, ctrl, target)) {
+                        addSwap(diag, ctrl, target, qubit_vertices);
+                        return it + 3;
+                    } else {
+                        addCnot(diag, ctrl, target, qubit_vertices);
+                    }
+
+                    break;
+                }
+                case qc::OpType::Z: {
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    addCnot(diag, ctrl, target, qubit_vertices);
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+
+                    break;
+                }
+
+                case qc::OpType::I: {
+                    break;
+                }
+
+                case qc::OpType::Phase: {
+                    const auto phase = PiRational(op->getParameter().front());
+                    addCphase(diag, phase, ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::T: {
+                    addCphase(diag, PiRational(1, 4), ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::S: {
+                    addCphase(diag, PiRational(1, 2), ctrl, target, qubit_vertices);
+                    break;
+                }
```

Comment:
these lines should be easy to cover in a test (probably even in an existing one).